### PR TITLE
Reject attempts to send event before privacy consent is given

### PR DESCRIPTION
**Pull request by richvdh**
Returns an [M_CONSENT_NOT_GIVEN error](https://github.com/matrix-org/matrix-doc/issues/1252) if consent is not yet given.

~~Builds on top of https://github.com/matrix-org/synapse/pull/3236.~~

### CODE REVIEW

File: synapse/handlers/message.py [hunk 1]
```diff
@@ -496,6 +508,79 @@ def create_event(self, requester, event_dict, token_id=None, txn_id=None,
...
+        Returns:
+            Deferred[bool]: true if the event can be sent without the user
+                consenting
+        """
+        # the only thing the user can do is join the server notices room.
+        if builder.type == EventTypes.Member:
+            membership = builder.content.get("membership", None)
+            if membership == Membership.JOIN:
+                return self._is_server_notices_room(builder.room_id)
+        return succeed(False)
```
Comment (erikjohnston) [inline synapse/handlers/message.py]: Do we not want to exempt AS users?

Comment (erikjohnston) [reply synapse/handlers/message.py]: Oh, yo do that later. I'm not sure I really get the difference between the checks here and the checks later

File: synapse/handlers/message.py [hunk 2]
```diff
@@ -496,6 +508,79 @@ def create_event(self, requester, event_dict, token_id=None, txn_id=None,
...
+                The user making the request
+
+        Returns:
+            Deferred[None]: returns normally if the user has consented or is
+                exempt
+
+        Raises:
+            ConsentNotGivenError: if the user has not given consent yet
+        """
+        logger.info("Consent for %s", requester.user)
```
Comment (erikjohnston) [inline synapse/handlers/message.py]: Should this really be info?

Comment (richvdh) [reply synapse/handlers/message.py]: no.

File: synapse/handlers/message.py (hunk 1 above)
Comment (richvdh) [reply synapse/handlers/message.py]: well, it's a difference between "does this *action* require consent to the privacy policy?" and "does this *user* need to consent to the privacy policy?".

the distinction means that it is meaningful to call `assert_accepted_privacy_policy` from the RoomCreationHandler, where we have already established that the action requires consent, but not whether the user needs to do so.

**Pull request merged**